### PR TITLE
ingress: prefer annotation over ingressClassName

### LINKED ISSUE #22340: ingress: incompatible with cert-manager ACME DNS-01
### What is happening?

With cilium `1.13.0-rc2`, set the default ingress class to `shared` (this functionality was implemented in #21386). Create a basic `ClusterIssuer` for the Let's Encrypt ACME HTTP01 challenge. Create an Ingress with the appropriate annotation and `spec.tls` setup properly. Observe as cert manager spins in a loop and creates hundreds of ingresses.

Screenshot 2022-11-23 at 5 16 06 PM

### Why is this happening?

This one took me a while to track down!

This happens because **cert-manager does not set ingressClassName**: see https://github.com/cert-manager/cert-manager/commit/65902d57a32fb9cc76bfae24a81400f1560cfc78

If the Kubernetes admission controller sees something without ingressClassName, and without the legacy ingressclass annotation, it will apply the default one ([relevant source lines from kubernetes/kubernetes](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/network/defaultingressclass/admission.go#L85)), and **in the process, will strip ownerReference**. This causes cert-manager to believe that it doesn't own the ingress, and so it creates a new one, where the ownerReference gets stripped, which causes cert-... you get the point. A control loop creating infinite objects makes my cluster very sad.

### Possible Fixes

- respect the legacy `kubernetes.io/ingress.class` annotation, if `ingressClassName` is not set (apparently every other ingress controller works this way, since cert-manager + HTTP01 ACME seems like a very popular use case)
- decide we won't fix this in cilium, and cert-manager needs to have an option to set `ingressClassName` (which is reasonable)

cc @sayboras 

### Cilium Version

1.13.0-rc2

### Kernel Version

Linux <hostname> 6.0.5-200.fc36.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Oct 26 15:55:21 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.3", GitCommit:"434bfd82814af038ad94d62ebe59b133fcb50506", GitTreeState:"clean", BuildDate:"2022-10-12T10:49:09Z", GoVersion:"go1.19.2", Compiler:"gc", Platform:"linux/amd64"}

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by nikhiljha**
- [x] All code is covered by unit and/or runtime tests where feasible.

The Kubernetes API documentation says that the ingressClassName field "replaces the deprecated `kubernetes.io/ingress.class` annotation. For backwards compatibility, when that annotation is set, it must be given precedence over this field." Other Kubernetes software relies on this, for example cert-manager and the Hashicorp Vault Helm chart.

This patch makes the Cilium ingress controller conform to the spec, by preferring the annotation if it is set, and then falling back to ingressClassName.

Fixes: #22340

Signed-off-by: Nikhil Jha <hi@nikhiljha.com>

```release-note
Cilium now prefers the `kubernetes.io/ingress.class` annotation over the `spec.ingressClassName` field when handling a new Ingress
```

### CODE REVIEW

**Comment by youngnick** [comment]
The change looks okay to me, but we definitely need some tests here, I think.

**Comment by nikhiljha** [comment]
Yep! I made the PR early but was planning on adding tests (left that box unchecked). I just added a copy of the basic Ingress test, but with the annotation rather than ingressClassName, which should be sufficient(?)

**Comment by youngnick** [comment]
Maybe a test that the annotation value overrides the field correctly? The current test makes sense, but we should test that extra type of behavior.

**Review by nebril — APPROVED** [review]

**Comment by nikhiljha** [comment]
This should probably be `needs-backport/1.13` as well, but I can't add tags 😁

**Comment by christarazi** [comment]
@nikhiljha Typically, if we label a PR with `release-note/bug`, then we'd need to consider how we might want the release note to be read by users. Currently, the PR title will be taken unless there's a `release-note`  in the description, see step 12 in [docs](https://docs.cilium.io/en/v1.12/contributing/development/contributing_guide/#submit-pr). As it stands, the PR title is not readable enough to users, i.e. it doesn't describe what the impact of the prior behavior is and what this PR fixes. This is so that users can read the changelog and understand what has been fixed in newer versions.

What's not clear to me is did the behavior prior to this PR exist in an already released version of Cilium (besides the RC)? If so, then I'd agree with `release-note/bug`, otherwise `release-note/misc` would be more appropriate. 

cc @youngnick

**Comment by nikhiljha** [comment]
No previous version of Cilium had this behavior, although it is a bug in the sense that it previously did not comply with the Ingress spec (which you would expect from an ingress controller). I think either way would make sense to me.

I updated the release note to be more descriptive-- let me know if this works!

> Cilium now prefers the `kubernetes.io/ingress.class` annotation over the `spec.ingressClassName` field when determining whether or not to handle an Ingress, as is mentioned in the Ingress spec.

One thought I had was failing if `kubernetes.io/ingress.class` and `spec.ingressClassName` are both set, but set to different things -- because that could cause some bad interaction between Cilium and other ingress controllers (although afaict all the popular ones handle this correctly). This is probably fine either way.

I also probably need to add this to the docs. I'll update that as well.

File: operator/pkg/ingress/doc.go
Comment (youngnick) [inline operator/pkg/ingress/doc.go]: Small non-blocking nit:
```suggestion
// or spec.ingressClassName set to "cilium" are managed and processed by the
```

**Review by youngnick — APPROVED** [review]
This LGTM from a ingress-spec POV, the extra test is great.

It's not the end of the world if two ingress controllers pick up the same ingress, because they generally have separate data paths. So I don't think this is worth holding up on that count.

FWIW, I think that not supporting the annotation per the spec previously was kind of a bug. But either release note category is fine with me.

**Review by joestringer — APPROVED** [review]

**Comment by joestringer** [comment]
> FWIW, I think that not supporting the annotation per the spec previously was kind of a bug. But either release note category is fine with me.

Do you think it's worth backporting this? If it's more of a bug then it might be nice to at least get into 1.13.

**Comment by nikhiljha** [comment]
Is there anything else needed on this one? I just rebased the changes on latest `master` branch.

**Comment by youngnick** [comment]
Looks like there's some failing tests, and I think this will need another rebase, but I think there's all the required approvals now. Maybe @joestringer can confirm there?

**Comment by joestringer** [comment]
Yep, looks like all of the required review approvals are in, it's mainly just testing that is blocking this PR. I'm not familiar with the latest state of the tests, but if the ingress conformance test is failing then there's a fair likelihood that the changes are related to that failure.

**Comment by aanm** [comment]
Ingress tests passed. Merging

**Pull request merged**
